bin/internal/update_dart_sdk.sh

```diff
@@ -56,20 +56,44 @@ if [ ! -f "$ENGINE_STAMP" ] || [ "$ENGINE_VERSION" != `cat "$ENGINE_STAMP"` ]; t
     echo
     exit 1
   }
-  >&2 echo "Downloading Dart SDK from Flutter engine $ENGINE_VERSION..."
 
-  # On x64 stdout is "uname -m: x86_64"
-  # On arm64 stdout is "uname -m: aarch64, arm64_v8a"
-  case "$(uname -m)" in
-    x86_64)
-      ARCH="x64"
-      ;;
-    *)
-      ARCH="arm64"
-      ;;
-  esac
+  OS="$(uname -s)"
+  # `uname -m` may be running in Rosetta mode, instead query sysctl
+  if [ "$OS" = 'Darwin' ]; then
+    # Allow non-zero exit so we can do control flow
+    set +e
+    # -n means only print value, not key
+    QUERY="sysctl -n hw.optional.arm64"
+    # Do not wrap $QUERY in double quotes, otherwise the args will be treated as
+    # part of the command
+    QUERY_RESULT=$($QUERY 2>/dev/null)
+    if [ $? -eq 1 ]; then
+      # If this command fails, we're certainly not on ARM
+      ARCH='x64'
+    elif [ "$QUERY_RESULT" = '0' ]; then
+      # If this returns 0, we are also not on ARM
+      ARCH='x64'
+    elif [ "$QUERY_RESULT" = '1' ]; then
+      ARCH='arm64'
+    else
+      >&2 echo "$QUERY returned unexpected output: $QUERY_RESULT"
```

Comment:
Consider single quotes around `$QUERY` and `$QUERY_RESULT`.

Reply:
done